File: client-layer/src/app/services/driver.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../enviroment';
import { DriverRegistrationRequest, DriverResponse } from '../models/driver.model';
import { RideStatus } from '../models/enums';

export interface DriverAssignedRide {
  rideId: number;
  startLocation: string;
  endLocation: string;
  startLatitude: number;
  startLongitude: number;
  endLatitude: number;
  endLongitude: number;
  status: RideStatus;
  scheduledAt?: string;
  estimatedPrice: number;
  passengerNames: string[];
  vehicleType: string;
}

@Injectable({
  providedIn: 'root'
})
export class DriverService {
  private apiUrl = `${environment.apiUrl}/drivers`;

  constructor(private http: HttpClient) {}

  // 2.2.3 Register Driver (Admin only)
  registerDriver(data: DriverRegistrationRequest): Observable<DriverResponse> {
    return this.http.post<DriverResponse>(this.apiUrl, data);
  }

  // Get assigned rides for driver
  getAssignedRides(driverId: number): Observable<DriverAssignedRide[]> {
    return this.http.get<DriverAssignedRide[]>(`${this.apiUrl}/${driverId}/assigned-rides`);
  }

  // Accept a ride
  acceptRide(driverId: number, rideId: number): Observable<any> {
    return this.http.post(`${this.apiUrl}/${driverId}/rides/${rideId}/accept`, {});
  }

  // Start a ride
  startRide(driverId: number, rideId: number): Observable<any> {
    return this.http.post(`${this.apiUrl}/${driverId}/rides/${rideId}/start`, {});
  }
}

File: mobile-application/app/src/main/java/com/example/mobile_application/service/DriverService.java
package com.example.mobile_application.service;

import com.example.mobile_application.dto.DriverAssignedRideDTO;
import com.example.mobile_application.dto.DriverRegistrationRequestDTO;
import com.example.mobile_application.dto.DriverResponseDTO;

import java.util.List;

import retrofit2.Call;
import retrofit2.http.Body;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.Path;

public interface DriverService {

    @POST("api/drivers")
    Call<DriverResponseDTO> registerDriver(@Body DriverRegistrationRequestDTO request);

    @GET("api/drivers/{driverId}/assigned-rides")
    Call<List<DriverAssignedRideDTO>> getAssignedRides(@Path("driverId") Long driverId);

    @POST("api/drivers/{driverId}/rides/{rideId}/start")
    Call<Void> startRide(@Path("driverId") Long driverId, @Path("rideId") Long rideId);
}
